tests/src/sanity.rs

```diff
@@ -1067,35 +1089,77 @@ mod tests {
+    #[timeout(Duration::from_secs(240))]
+    pub async fn test_outgoing_traffic_udp_with_connect(
+        #[future] udp_logger_service: KubeService,
+        #[future] service: KubeService,
+        #[future] kube_client: Client,
+    ) {
+        let internal_service = udp_logger_service.await; // Only reachable from withing the cluster.
+        let target_service = service.await; // Impersonate a pod of this service, to reach internal.
+        let kube_client = kube_client.await;
+        let pod_api: Api<Pod> = Api::namespaced(kube_client.clone(), &internal_service.namespace);
+        let lp = LogParams {
+            container: Some(String::from(CONTAINER_NAME)),
+            follow: false,
+            limit_bytes: None,
+            pretty: false,
+            previous: false,
+            since_seconds: None,
+            tail_lines: None,
+            timestamps: false,
+        };
 
         let node_command = vec![
             "node",
-            "node-e2e/outgoing/test_outgoing_traffic_udp_client.mjs",
-            &port,
+            "node-e2e/outgoing/test_outgoing_traffic_udp_client_with_connect.mjs",
+            "31415",
+            // Reaching service by only service name is only possible from within the cluster.
+            &internal_service.name,
         ];
-        let mut process = run(node_command, &service.pod_name, None, None).await;
 
-        // Listen for UDP message from mirrorded app.
-        let mut buf = [0; 27];
-        let amt = socket.recv(&mut buf).unwrap();
-        assert_eq!(amt, 27);
-        assert_eq!(buf, "Can I pass the test please?".as_ref()); // Sure you can.
+        // Meta-test: verify that the application cannot reach the internal service without
+        // mirrord forwarding outgoing UDP traffic via the target pod.
+        // If this verification fails, the test itself is invalid.
+        let mirrord_no_outgoing = vec!["--no-outgoing"];
+        let mut process = run(
+            node_command.clone(),
+            &target_service.pod_name,
+            Some(&target_service.namespace),
+            Some(mirrord_no_outgoing),
+        )
+        .await;
+        let res = process.child.wait().await.unwrap();
+        assert!(!res.success()); // Should fail because local process cannot reach service.
+        let logs = pod_api.logs(&internal_service.pod_name, &lp).await;
+        assert_eq!(logs.unwrap(), "");
 
+        // Run mirrord with outgoing enabled.
+        let mut process = run(
+            node_command,
+            &target_service.pod_name,
+            Some(&target_service.namespace),
+            None,
+        )
+        .await;
         let res = process.child.wait().await.unwrap();
         assert!(res.success());
         process.assert_stderr();
+
+        // Verify that the UDP message sent by the application reached the internal service.
+        sleep(Duration::from_secs(1)); // Let the server time to print the log.
```

Comment:
Can we just sample the logs until the message shows up, with a timeout?

Reply:
Thanks for that feedback. Changed it to wait for an item on a log stream :)
CI is still stuck though.